File: src/bubble.js
import { createHash, randomBytes, timingSafeEqual } from 'node:crypto';
import { spawn, spawnSync } from 'node:child_process';
import fs from 'node:fs';
import http from 'node:http';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { render as renderMarkdown } from './markdown.js';

const ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const PAGE = path.join(ROOT, 'ui', 'bubble.html');
const SHELL_SOURCE = path.join(ROOT, 'shell', 'SaynowPanel.swift');

const MIME = {
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.svg': 'image/svg+xml',
};

// A document can be wider than a sentence without becoming a window.
const WIDTH = 420;
const INITIAL_HEIGHT = 200;

const CACHE_DIR =
  process.env.SAYNOW_CACHE_DIR ||
  (process.platform === 'darwin'
    ? path.join(os.homedir(), 'Library', 'Caches', 'saynow')
    : path.join(process.env.XDG_CACHE_HOME || path.join(os.homedir(), '.cache'), 'saynow'));

/**
 * Show the bubble and wait for the user.
 *
 * Resolves with {reason: 'reply', text} when they answer, or
 * {reason: 'dismiss'} when it times out or they wave it away.
 */
export async function showBubble({
  text,
  ask = false,
  rate,
  from,
  document,
  dismissMs = 5000,
  onStop,
  speech,
  audio,
  audioType,
  chunks,
  timeoutMs = 120_000,
} = {}) {
  // When the page plays the audio itself, the highlight is driven by the
  // element's own currentTime — exact by construction, rather than a
  // words-per-minute guess that starts whenever the page happened to load.
  // Chunks are synthesised ahead of the page asking for them, so playback of
  // one overlaps rendering of the next and speech starts almost immediately.
  const pending = chunks ? new Map() : null;
  if (chunks) prefetch(chunks, pending, 0);

  const state = {
    text,
    ask,
    // Who is speaking. Shown in the header so the sentence itself does not
    // have to spend words introducing the agent every single time.
    from: from || null,
    // Rendered here rather than in the page: the renderer escapes its input
    // and emits only tags it built, so the page never parses agent text.
    html: document
      ? renderMarkdown(document.markdown, {
          asset: (src) =>
            /^[a-z][\w+.-]*:/i.test(src)
              ? null // an unknown scheme is dropped rather than guessed at
              : `/asset?t=${token}&p=${encodeURIComponent(src)}`,
        })
      : null,
    rate: rate || 175,
    dismissMs,
    hasAudio: Boolean(audio),
    chunks: chunks?.map((c) => ({ firstWord: c.firstWord, wordCount: c.wordCount })) ?? null,
  };
  const page = renderPage(state);

  /**
   * A secret for this bubble only.
   *
   * The server listens on loopback with no other authentication, and POST
   * /reply is what an agent reads as your answer — so anything able to reach
   * the port could answer on your behalf and the agent would act on it. The
   * page is served the token and hands it back; nothing else knows it.
   */
  const token = randomBytes(24).toString('hex');

  const authorised = (req) => {
    const supplied =
      req.headers['x-saynow-token'] ||
      new URL(req.url ?? '/', 'http://localhost').searchParams.get('t') ||
      '';
    const a = Buffer.from(String(supplied));
    const b = Buffer.from(token);
    return a.length === b.length && timingSafeEqual(a, b);
  };

  let settle;
  const answered = new Promise((resolve) => (settle = resolve));

  // The word animation is only an estimate of pace. The dismiss countdown has
  // to start when the audio genuinely stops, or the bubble can vanish
  // mid-sentence on a slow voice.
  const listeners = new Set();
  const announceSpeechEnded = () => {
    for (const res of listeners) res.write('event: ended\ndata: {}\n\n');
  };

  const server = http.createServer((req, res) => {
    if (!authorised(req)) {
      res.writeHead(403).end();
      return;
    }

    const route = (req.url ?? '/').split('?')[0];
    // /audio/<n> blocks until that chunk is rendered, which is what lets the
    // page ask for the next one before it exists.
    const chunkMatch = req.method === 'GET' && /^\/audio\/(\d+)$/.exec(route);
    if (chunkMatch && chunks) {
      const index = Number(chunkMatch[1]);
      prefetch(chunks, pending, index + 1);
      Promise.resolve(pending.get(index))
        .then((rendered) => {
          if (!rendered?.audio) {
            res.writeHead(503).end();
            return;
          }
          res.writeHead(200, {
            'Content-Type': rendered.ext === 'mp3' ? 'audio/mpeg' : 'audio/wav',
            'Content-Length': rendered.audio.length,
            'Cache-Control': 'no-store',
          });
          res.end(rendered.audio);
        })
        .catch(() => res.writeHead(503).end());
      return;
    }

    // Images referenced by the document, resolved against its own directory
    // and refused if they climb out of it.
    if (req.method === 'GET' && route === '/asset' && document) {
      const rel = new URL(req.url, 'http://localhost').searchParams.get('p') ?? '';
      const full = path.resolve(document.dir, rel);
      if (!full.startsWith(document.dir + path.sep)) {
        res.writeHead(403).end();
        return;
      }
      try {
        const body = fs.readFileSync(full);
        res.writeHead(200, {
          'Content-Type': MIME[path.extname(full).toLowerCase()] ?? 'application/octet-stream',
          'Content-Length': body.length,
        });
        res.end(body);
      } catch {
        res.writeHead(404).end();
      }
      return;
    }

    if (req.method === 'GET' && route === '/audio' && audio) {
      res.writeHead(200, {
        'Content-Type': audioType || 'audio/wav',
        'Content-Length': audio.length,
        'Accept-Ranges': 'none',
        'Cache-Control': 'no-store',
      });
      res.end(audio);
      return;
    }

    if (req.method === 'GET' && route === '/events') {
      res.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-store',
        Connection: 'keep-alive',
      });
      res.write('retry: 1000\n\n');
      listeners.add(res);
      req.on('close', () => listeners.delete(res));
      return;
    }

    if (req.method === 'GET') {
      res.writeHead(200, {
        'Content-Type': 'text/html; charset=utf-8',
        'Cache-Control': 'no-store',
      });
      res.end(page);
      return;
    }

    readJson(req).then((body) => {
      res.writeHead(204).end();
      switch (route) {
        case '/stop':
          onStop?.();
          break;
        case '/reply':
          settle({ reason: 'reply', text: String(body.text ?? '').trim() });
          break;
        case '/dismiss':
          settle({ reason: 'dismiss' });
          break;
        case '/close':
          settle({ reason: 'dismiss' });
          break;
      }
    });
  });

  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const url = `http://127.0.0.1:${server.address().port}/?t=${token}`;

  const window = openWindow(url);
  const guard = setTimeout(() => settle({ reason: 'dismiss' }), timeoutMs);

  // Resolves when the audio actually finishes, however long it really took.
  Promise.resolve(speech)
    .catch(() => {})
    .then(announceSpeechEnded);

  try {
    return await answered;
  } finally {
    clearTimeout(guard);
    for (const res of listeners) res.end();
    window.close();
    server.close();
  }
}

/** Keep a small lookahead rendering, so the page rarely has to wait. */
const LOOKAHEAD = 2;

function prefetch(chunks, pending, from) {
  for (let i = from; i < Math.min(chunks.length, from + LOOKAHEAD); i += 1) {
    if (!pending.has(i)) pending.set(i, chunks[i].render());
  }
}

function renderPage(state) {
  const html = fs.readFileSync(PAGE, 'utf8');
  const injected = `<script>window.__SAYNOW__ = ${JSON.stringify(state).replace(
    /</g,
    '\\u003c',
  )};</script>`;
  return html.replace('<!--SAYNOW_STATE-->', injected);
}

function readJson(req) {
  return new Promise((resolve) => {
    const chunks = [];
    req.on('data', (c) => chunks.push(c));
    req.on('end', () => {
      try {
        resolve(JSON.parse(Buffer.concat(chunks).toString() || '{}'));
      } catch {
        resolve({});
      }
    });
    req.on('error', () => resolve({}));
  });
}

/* ---- window shells ------------------------------------------------------ */

function openWindow(url) {
  return nativePanel(url) || browserWindow(url) || noWindow();
}

/**
 * The preferred shell: a borderless transparent NSPanel. Compiled on first use
 * and cached, keyed by a hash of the source so an upgrade rebuilds it.
 */
function nativePanel(url) {
  if (process.platform !== 'darwin') return null;
  if (!fs.existsSync(SHELL_SOURCE)) return null;
  if (spawnSync('which', ['swiftc'], { stdio: 'ignore' }).status !== 0) return null;

  const source = fs.readFileSync(SHELL_SOURCE);
  const digest = createHash('sha256').update(source).digest('hex').slice(0, 12);
  const binary = path.join(CACHE_DIR, `panel-${digest}`);

  if (!fs.existsSync(binary)) {
    fs.mkdirSync(CACHE_DIR, { recursive: true });
    const build = spawnSync('swiftc', ['-O', '-o', binary, SHELL_SOURCE], {
      stdio: 'ignore',
      timeout: 120_000,
    });
    if (build.status !== 0) return null;
    // A stale build from an older version is just wasted disk; clear it out.
    for (const entry of fs.readdirSync(CACHE_DIR)) {
      if (entry.startsWith('panel-') && entry !== `panel-${digest}`) {
        fs.rmSync(path.join(CACHE_DIR, entry), { force: true });
      }
    }
  }

  const child = spawn(binary, [url, String(WIDTH), String(INITIAL_HEIGHT)], {
    stdio: ['pipe', 'ignore', 'ignore'],
  });

  // The panel exits when this pipe closes, so it can never outlive us.
  return { close: () => child.kill() };
}

const BROWSERS = {
  darwin: [
    '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
    '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser',
    '/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge',
    '/Applications/Chromium.app/Contents/MacOS/Chromium',
  ],
  linux: ['google-chrome', 'chromium', 'chromium-browser', 'microsoft-edge'],
  win32: [
    'C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe',
    'C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe',
  ],
};

/** Fallback shell: a Chromium app window. Same page, but the OS draws a frame. */
function browserWindow(url) {
  const candidates = BROWSERS[process.platform] || [];
  const browser = candidates.find(
    (c) =>
      (c.includes('/') || c.includes('\\')
        ? fs.existsSync(c)
        : spawnSync('which', [c], { stdio: 'ignore' }).status === 0),
  );
  if (!browser) return null;

  const { width, height } = screenSize();
  const child = spawn(
    browser,
    [
      `--app=${url}`,
      `--window-size=${WIDTH},${INITIAL_HEIGHT + 40}`,
      `--window-position=${Math.max(0, width - WIDTH - 24)},${Math.max(0, height - INITIAL_HEIGHT - 120)}`,
      `--user-data-dir=${path.join(CACHE_DIR, 'browser-profile')}`,
      '--no-first-run',
      '--no-default-browser-check',
      // The page plays the speech itself; without this it is blocked as
      // unsolicited autoplay and the bubble sits silent.
      '--autoplay-policy=no-user-gesture-required',
    ],
    { stdio: 'ignore', detached: false },
  );

  return { close: () => child.kill() };
}

function screenSize() {
  if (process.platform === 'darwin') {
    const out = spawnSync(
      'osascript',
      ['-e', 'tell application "Finder" to get bounds of window of desktop'],
      { encoding: 'utf8' },
    );
    const parts = (out.stdout || '').trim().split(', ').map(Number);
    if (parts.length === 4 && parts[2] > 0) return { width: parts[2], height: parts[3] };
  }
  return { width: 1440, height: 900 };
}

/** No shell available — speak-only, so --ask degrades to a plain announcement. */
function noWindow() {
  return { close: () => {} };
}

export function canShowBubble() {
  if (process.platform === 'darwin' && spawnSync('which', ['swiftc'], { stdio: 'ignore' }).status === 0) {
    return 'panel';
  }
  const candidates = BROWSERS[process.platform] || [];
  const found = candidates.some((c) =>
    c.includes('/') || c.includes('\\')
      ? fs.existsSync(c)
      : spawnSync('which', [c], { stdio: 'ignore' }).status === 0,
  );
  return found ? 'browser' : null;
}

File: python/src/saynow/bubble.py
"""The floating bubble, for the pip build.

A port of src/bubble.js. The page itself is shared verbatim — bubble.html is
copied into this package rather than reimplemented, so the two builds cannot
drift into looking or behaving differently.
"""

from __future__ import annotations

import hashlib
import json
import mimetypes
import os
import re
import secrets
import shutil
import socketserver
import subprocess
import sys
import threading
import time
from http.server import BaseHTTPRequestHandler
from pathlib import Path
from typing import Any, Dict, Optional
from urllib.parse import parse_qs, quote, urlparse

from .markdown import render as render_markdown

WIDTH = 420
INITIAL_HEIGHT = 200

BROWSERS = {
    "darwin": [
        "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
        "/Applications/Brave Browser.app/Contents/MacOS/Brave Browser",
        "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge",
    ],
    "linux": ["google-chrome", "chromium", "chromium-browser", "microsoft-edge"],
}


def cache_dir() -> Path:
    override = os.environ.get("SAYNOW_CACHE_DIR")
    if override:
        return Path(override).expanduser()
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Caches" / "saynow"
    base = os.environ.get("XDG_CACHE_HOME") or (Path.home() / ".cache")
    return Path(base) / "saynow"


def available() -> Optional[str]:
    """Which window shell can be used, if any."""
    if sys.platform == "darwin" and shutil.which("swiftc"):
        return "panel"
    for candidate in BROWSERS.get(sys.platform, []):
        if Path(candidate).exists() or shutil.which(candidate):
            return "browser"
    return None


def resolve_asset(directory: Path, relative: str) -> Optional[Path]:
    """Resolve a document-relative path, or None if it climbs out.

    A document is shown with the reader's own files reachable behind it, so
    this is the only thing stopping "../../../.ssh/id_rsa" from being served
    to a page whose content an agent wrote.
    """
    root = Path(directory).resolve()
    full = (root / relative).resolve()
    if full == root or not str(full).startswith(str(root) + os.sep):
        return None
    return full


def _page(state: Dict[str, Any]) -> bytes:
    html = (Path(__file__).parent / "bubble.html").read_text(encoding="utf-8")
    injected = "<script>window.__SAYNOW__ = {};</script>".format(
        json.dumps(state).replace("<", "\\u003c")
    )
    return html.replace("<!--SAYNOW_STATE-->", injected).encode("utf-8")


class _Server(socketserver.ThreadingTCPServer):
    daemon_threads = True
    allow_reuse_address = True


def show_bubble(
    text: str,
    ask: bool = False,
    sender: Optional[str] = None,
    audio: Optional[bytes] = None,
    audio_ext: str = "wav",
    document: Optional[Dict[str, Any]] = None,
    rate: Optional[float] = None,
    dismiss_ms: Optional[int] = None,
    timeout: float = 120.0,
) -> Dict[str, Any]:
    """Show the bubble and wait. Returns {"reason": "reply"|"dismiss", "text": ...}."""

    # A secret for this bubble alone. POST /reply is what the caller reads back
    # as the user's answer, so anything able to reach the port could otherwise
    # answer on their behalf. Declared before the asset callback that closes
    # over it — in the npm build a reference from above threw for documents
    # carrying a relative image and left every other one working.
    token = secrets.token_hex(24)
    document_dir = Path(document["dir"]) if document and document.get("dir") else None

    def asset(src: str) -> Optional[str]:
        # An unknown scheme is dropped rather than guessed at.
        if re.match(r"^[a-z][\w+.-]*:", src, re.I):
            return None
        return f"/asset?t={token}&p={quote(src, safe='')}"

    state = {
        "text": text,
        "ask": ask,
        "from": sender or None,
        "rate": rate or 175,
        # A question needs time to be read and answered, so it holds four
        # times longer than a statement you only have to hear.
        "dismissMs": dismiss_ms or (20000 if ask else 5000),
        "hasAudio": bool(audio),
        # Rendered here rather than in the page: the renderer escapes its input
        # and emits only tags it built, so the page never parses agent text.
        "html": render_markdown(document["markdown"], asset=asset) if document else None,
        "chunks": None,
    }
    page = _page(state)

    settled: Dict[str, Any] = {}
    done = threading.Event()

    def settle(reason: str, body: Optional[Dict[str, Any]] = None) -> None:
        if not done.is_set():
            settled.update({"reason": reason, "text": (body or {}).get("text", "")})
            done.set()

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *_args):  # noqa: D401 - silence the default logging
            pass

        def _authorised(self) -> bool:
            supplied = self.headers.get("X-Saynow-Token") or (
                parse_qs(urlparse(self.path).query).get("t", [""])[0]
            )
            return secrets.compare_digest(str(supplied), token)

        def _send(self, code: int, body: bytes = b"", content_type: str = "text/plain"):
            self.send_response(code)
            self.send_header("Content-Type", content_type)
            self.send_header("Content-Length", str(len(body)))
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            if body:
                self.wfile.write(body)

        def do_GET(self):  # noqa: N802 - required by BaseHTTPRequestHandler
            if not self._authorised():
                self._send(403)
                return
            route = urlparse(self.path).path

            if route == "/audio" and audio:
                self._send(
                    200, audio, "audio/mpeg" if audio_ext == "mp3" else "audio/wav"
                )
                return

            if route == "/asset" and document_dir:
                rel = parse_qs(urlparse(self.path).query).get("p", [""])[0]
                full = resolve_asset(document_dir, rel)
                if full is None:
                    self._send(403)
                    return
                try:
                    self._send(
                        200,
                        full.read_bytes(),
                        mimetypes.guess_type(str(full))[0] or "application/octet-stream",
                    )
                except OSError:
                    self._send(404)
                return

            if route == "/events":
                # Held open but silent: with audio present the page drives the
                # transcript from the audio clock and never needs this.
                self.send_response(200)
                self.send_header("Content-Type", "text/event-stream")
                self.end_headers()
                try:
                    while not done.is_set():
                        time.sleep(0.2)
                except OSError:
                    pass
                return

            self._send(200, page, "text/html; charset=utf-8")

        def do_POST(self):  # noqa: N802
            if not self._authorised():
                self._send(403)
                return
            route = urlparse(self.path).path
            length = int(self.headers.get("Content-Length") or 0)
            try:
                body = json.loads(self.rfile.read(length) or b"{}")
            except (json.JSONDecodeError, ValueError):
                body = {}
            self._send(204)

            if route == "/reply":
                settle("reply", body)
            elif route in ("/dismiss", "/close"):
                settle("dismiss")

    server = _Server(("127.0.0.1", 0), Handler)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()

    url = f"http://127.0.0.1:{port}/?t={token}"
    window = _open_window(url)

    try:
        done.wait(timeout=timeout)
        return settled or {"reason": "dismiss", "text": ""}
    finally:
        if window:
            window.terminate()
        server.shutdown()
        server.server_close()


def _open_window(url: str):
    """Prefer the native panel; fall back to a Chromium app window."""
    # Tests need the server without putting a window on someone's screen.
    if os.environ.get("SAYNOW_NO_WINDOW"):
        return None
    if sys.platform == "darwin" and shutil.which("swiftc"):
        source = Path(__file__).parent / "SaynowPanel.swift"
        digest = hashlib.sha256(source.read_bytes()).hexdigest()[:12]
        binary = cache_dir() / f"panel-{digest}"

        if not binary.exists():
            cache_dir().mkdir(parents=True, exist_ok=True)
            built = subprocess.run(
                ["swiftc", "-O", "-o", str(binary), str(source)],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                timeout=120,
            )
            if built.returncode != 0:
                return _open_browser(url)

        # The panel exits when this pipe closes, so it cannot outlive us.
        return subprocess.Popen(
            [str(binary), url, str(WIDTH), str(INITIAL_HEIGHT)],
            stdin=subprocess.PIPE,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )

    return _open_browser(url)


def _open_browser(url: str):
    for candidate in BROWSERS.get(sys.platform, []):
        path = candidate if Path(candidate).exists() else shutil.which(candidate)
        if not path:
            continue
        return subprocess.Popen(
            [
                path,
                f"--app={url}",
                f"--window-size={WIDTH},{INITIAL_HEIGHT + 40}",
                f"--user-data-dir={cache_dir() / 'browser-profile'}",
                "--no-first-run",
                "--no-default-browser-check",
                "--autoplay-policy=no-user-gesture-required",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    return None
